[quant] EqualizationQConfig to distinguish input/output activations

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59953 [quant] Equalization Observer modifications
* #59747 [quant] Input Weight Equalization - prepare modifications
* **#59739 [quant] EqualizationQConfig to distinguish input/output activations**

Summary: Created an EqualizationQConfig specifically for equalization.
This inherits from QConfig and is used to distinguish between inserting
an input observer with an output observer. Since the output observer
field is included in the EqualizationQConfig, we no longer need an
output observer field in the _InputEqualizationObserver

Test Plan: compiles

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29135298](https://our.internmc.facebook.com/intern/diff/D29135298)

angelayi committed:
[quant] EqualizationQConfig to distinguish input/output activations

Summary: Created an EqualizationQConfig specifically for equalization.
This inherits from QConfig and is used to distinguish between inserting
an input observer with an output observer. Since the output observer
field is included in the EqualizationQConfig, we no longer need an
output observer field in the _InputEqualizationObserver

Test Plan: compiles

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
looks good. I think we may want to eventually move this into the qconfig.py file when it is more stable. For now seems like this is a good place as we develop this feature.

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -229,3 +225,41 @@ def calculate_equalization_scale(input_obs: _InputEqualizationObserver,
+                                    weight=_WeightEqualizationObserver.with_args(dtype=torch.qint8))
+    """
+    def __new__(cls, input_activation=torch.nn.Identity, output_activation=torch.nn.Identity, weight=torch.nn.Identity):
+        # catch common mistakes
```
nit: remove this comment?

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -229,3 +225,41 @@ def calculate_equalization_scale(input_obs: _InputEqualizationObserver,
+            raise ValueError("EqualizationQConfig received observer instance, please pass observer class instead. " +
+                             "Use MyObserver.with_args(x=1) to override arguments to constructor if needed")
+        self = super(EqualizationQConfig, cls).__new__(cls, output_activation, weight)
+        self.input_activation = input_activation
```
Should we also assert on the type of input_activation and weight? For equalization we should only support _InputEqualizationObserver and similar for weight

angelayi replied:
I'm not sure how to check the type of the input_activation/weights as they are `functool.partial`s. I could do `isinstance(input_activation(), _InputEqualizationObserver)`, but that would create an `input_activation` instance.

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -229,3 +225,41 @@ def calculate_equalization_scale(input_obs: _InputEqualizationObserver,
+class EqualizationQConfig(QConfig):
+    """
+    Describes how to quantize a layer or a part of the network specifically for
+    equalization by providing settings (observer classes) for inputs, outputs,
```
Let's say input-weight equalization in the description to be more specific

angelayi committed:
Update on "[quant] EqualizationQConfig to distinguish input/output activations"


Summary: Created an EqualizationQConfig specifically for equalization.
This inherits from QConfig and is used to distinguish between inserting
an input observer with an output observer. Since the output observer
field is included in the EqualizationQConfig, we no longer need an
output observer field in the _InputEqualizationObserver

Test Plan: compiles

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] EqualizationQConfig to distinguish input/output activations"


Summary: Created an EqualizationQConfig specifically for equalization.
This inherits from QConfig and is used to distinguish between inserting
an input observer with an output observer. Since the output observer
field is included in the EqualizationQConfig, we no longer need an
output observer field in the _InputEqualizationObserver

Test Plan: compiles

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] EqualizationQConfig to distinguish input/output activations"


Summary: Created an EqualizationQConfig specifically for equalization.
This inherits from QConfig and is used to distinguish between inserting
an input observer with an output observer. Since the output observer
field is included in the EqualizationQConfig, we no longer need an
output observer field in the _InputEqualizationObserver

Test Plan: compiles

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] EqualizationQConfig to distinguish input/output activations"


Summary: Created an EqualizationQConfig specifically for equalization.
This inherits from QConfig and is used to distinguish between inserting
an input observer with an output observer. Since the output observer
field is included in the EqualizationQConfig, we no longer need an
output observer field in the _InputEqualizationObserver

Test Plan: compiles

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29135298](https://our.internmc.facebook.com/intern/diff/D29135298)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29135298).

angelayi committed:
Update on "[quant] EqualizationQConfig to distinguish input/output activations"


Summary: Created an EqualizationQConfig specifically for equalization.
This inherits from QConfig and is used to distinguish between inserting
an input observer with an output observer. Since the output observer
field is included in the EqualizationQConfig, we no longer need an
output observer field in the _InputEqualizationObserver

Test Plan: compiles

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29135298](https://our.internmc.facebook.com/intern/diff/D29135298)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29135298).
